src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs

```diff
@@ -1340,6 +1341,43 @@ public bool WaitForExit(int milliseconds)
             return exited;
         }
 
+        /// <summary>
+        /// Instructs the Process component to wait for the associated process to exit, or
+        /// for the <paramref name="cancellationToken"/> to be canceled.
+        /// </summary>
+        public async Task WaitForExitAsync(CancellationToken cancellationToken = default)
+        {
+            if (!Associated) { throw new InvalidOperationException(SR.NoAssociatedProcess); }
+            if (!EnableRaisingEvents) { throw new InvalidOperationException(SR.EnableRaisingEventsRequired); }
+
+            var tcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            void ExitedHandler(object sender, EventArgs e)
+            {
+                tcs.TrySetResult(null);
+            }
+
+            Exited += ExitedHandler;
+
+            try
+            {
+                if (HasExited)
+                {
+                    // Handle the race where the process exited prior to registering the event handler
+                    return;
+                }
+
+                using (cancellationToken.Register(() => tcs.TrySetCanceled()))
+                {
+                    await tcs.Task.ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                Exited -= ExitedHandler;
```

Comment:
This is going to allocate another delegate.  A better way to do this would be:
```C#
EventHandler handler = (s, e) => tcs.TrySetResult(null);
Exited += handler;
try
{
    ...
}
finally
{
    Exited -= handler;
}
```

Reply:
Thanks for this! I've updated the code.